src/BenchmarkDotNet/Diagnosers/EventPipeProfiler.cs

```diff
@@ -0,0 +1,235 @@
+        {
+            if (profile == null && (providers == null || !providers.Any()))
+            {
+                logger.WriteLine(LogKind.Info, "No profile or providers specified, defaulting to trace profile 'CpuSampling'");
+                profile = EventPipeProfile.CpuSampling;
+            }
+
+            if (providers != null)
+            {
+                eventPipeProviders.AddRange(providers);
+            }
+
+            if (profile != null)
+            {
+                if (EventPipeProfileMapper.DotNetRuntimeProfiles.TryGetValue(profile.Value, out var selectedProfile))
+                {
+                    var newProvidersFromProfile = selectedProfile.Where(p => !eventPipeProviders.Any(r => r.Name.Equals(p.Name)));
+                    eventPipeProviders.AddRange(newProvidersFromProfile);
+                }
+                else
+                {
+                    logger.WriteLine(LogKind.Error, $"Invalid profile name: {profile}");
+                }
+            }
+        }
+
+        private readonly Dictionary<BenchmarkCase, string> benchmarkToTraceFile = new Dictionary<BenchmarkCase, string>();
+        
+        private readonly List<EventPipeProvider> eventPipeProviders = new List<EventPipeProvider>
+        {
+            new EventPipeProvider("BenchmarkDotNet.EngineEventSource", EventLevel.Informational, long.MaxValue) // mandatory provider to enable Engine events
+        };
+
+        private static readonly string LogSeparator = new string('-', 20);
+
+        public static readonly EventPipeProfiler Default = new EventPipeProfiler();
+
+        private readonly LogCapture logger = new LogCapture();
+
+        private Task collectingTask;
+
+        public IEnumerable<string> Ids => new[] { nameof(EventPipeProfiler) };
+
+        public IEnumerable<IExporter> Exporters => Array.Empty<IExporter>();
+
+        public IEnumerable<IAnalyser> Analysers => Array.Empty<IAnalyser>();
+
+        public RunMode GetRunMode(BenchmarkCase benchmarkCase) => RunMode.ExtraRun;
+
+        public void Handle(HostSignal signal, DiagnoserActionParameters parameters)
+        {
+            if (signal != HostSignal.BeforeAnythingElse)
+                return;
+
+            var diagnosticsClient = new DiagnosticsClient(parameters.Process.Id);
+
+            EventPipeSession session;
+            try
+            {
+                session = diagnosticsClient.StartEventPipeSession(eventPipeProviders, true);
+            }
+            catch (DiagnosticsClientException e)
+            {
+                logger.WriteLine(LogKind.Error, $"Unable to start a tracing session: {e}");
+                return;
+            }
+            var fileName = ArtifactFileNameHelper.GetFilePath(parameters, DateTime.Now, "nettrace");
+            Path.GetDirectoryName(fileName).CreateIfNotExists();
+            benchmarkToTraceFile[parameters.BenchmarkCase] = fileName;
+
+            collectingTask = new Task(() =>
```

Comment:
nit: we have quite a deep nesting here. in a perfect world, we would move this logic to a separate method and avoid closure capture